src/lib/groovy/ModsDotGroovy.groovy

```diff
@@ -86,74 +153,134 @@ class ModsDotGroovy {
+                        }
+                        break
                 }
+                modData['authors'] = authorsString
+                mods.add(modData)
                 break
-        }
-        modData['authors'] = authorsString
+            case Platform.QUILT:
+                modInfo.dependencies?.each {
+                    ((this.data.computeIfAbsent('quilt_loader',{[:]}) as Map).computeIfAbsent('depends',{[]}) as List)
+                            .add(it.asQuiltMap())
+                }
+
+                if (modInfo.customProperties !== null && !modInfo.customProperties.isEmpty())
+                    this.data = merge(this.data, ['quilt_loader':modInfo.customProperties])
+
+                boolean otherQuiltModsExist = (this.data?.quilt_loader as Map)?.id !== null
+
+                final quiltModData = [:]
+                final quiltMetadata = [:]
+                if (!otherQuiltModsExist) {
+                    quiltModData['id'] = modInfo.modId
+                    quiltModData['version'] = modInfo.version
+                } else {
+                    quiltModData.computeIfAbsent('provides', {[]})
+                    List provides = quiltModData['provides'] as List
+                    provides.add(['id':modInfo.modId,'version':modInfo.version])
+                }
+                quiltMetadata['name'] = modInfo.displayName
+                quiltMetadata['contact'] = ["homepage":modInfo.displayUrl]
+                quiltMetadata['icon'] = modInfo.logoFile
+                quiltMetadata['description'] = modInfo.description
+                quiltModData['entrypoints'] = modInfo.entrypoints
 
-        mods.add(modData)
+                Map quiltContributors = [:]
+                modInfo.authors.each {
+                    quiltContributors[it] = "Author"
+                }
+                quiltMetadata['contributors'] = quiltContributors
+                quiltModData['metadata'] = quiltMetadata
+                this.data = merge(this.data, ["quilt_loader":quiltModData])
+        }
     }
 
     void sanitize() {
         sanitizeMap(data)
     }
 
     private static void sanitizeMap(Map data) {
-        final copy = Map.copyOf(data)
+        final copy = new LinkedHashMap(data) // cannot use Map.copyOf as we wish to remove null values
         copy.forEach { Object key, Object value ->
             if (value === null) data.remove(key)
             else if (value instanceof List) {
                 value.removeIf { it === null }
             } else if (value instanceof Map) {
                 sanitizeMap(value)
+            } else if (value instanceof GString) {
+                data[key] = value.toString()
             }
         }
     }
 
-    static ModsDotGroovy make(@DelegatesTo(value = ModsDotGroovy, strategy = DELEGATE_FIRST) Closure closure) {
+    private static Map merge(Map left, Map right) {
+        if (left === null) return right
+        if (right === null) return right
```

Comment:
```suggestion
        if (left === null || right === null) return right
```

Reply:
Oops, it's meant to return the non-null one